Refactor tzdump.rb (Ruby bin)

mexisme opened the pull request.

Commit:
Add Gemfile for required deps

Commit:
Add hash-bang

Commit:
DRY, more idiomatic

Commit:
`$stdout` should work as-well as copying the STDOUT FH into a new IO object

Commit:
`$stdout.write` --> puts_crlf()

Commit:
Extract some code from `dump_zone()` into `dump_period()`

Commit:
Add TODO

jskeet commented on ruby/tzdump.rb:
```diff
@@ -5,35 +7,50 @@
+        TZInfo::Timezone.all_identifiers.sort
+      else
+        ARGV[0]
+      end
```
Presumbly this could just be

```
tzs = ARGV.empty? ? TZInfo::Timezone.all_identifiers.sort : ARGV[0]
```

mexisme replied:
Both forms work, but I've been repeatedly lambasted against using the ternary operator, partly because of precedence mistakes, and partly because if...else is more readable.

jskeet replied:
Ick - it doesn't look more readable to me. Admittedly the `?` at the end of `ARGV.empty` makes it a bit more confusing than it would be otherwise. If Ruby makes precedence tricky here, that's fair enough. Not to my taste, but never mind.

mexisme replied:
No worries;  YMMV.

mexisme replied:
(Fixed)

jskeet commented on ruby/tzdump.rb:
```diff
@@ -5,35 +7,50 @@
 
-def dump_zone(console, id)
-  console.write "#{id}\r\n"
+def dump_period(period, zone)
```
I think I'll want to rename this - it sounds like it's just dumping a _single_ period, whereas actually it's dumping everything for the whole zone.

Also, the fact that it starts with a single period feels odd to me.

How about making it _genuinely_ print out the details for a single period (or actually a single transition), and return the next one - then putting the while loop in dump_zone instead?

mexisme replied:
> How about making it genuinely print out the details for a single period (or actually a single transition), and return the next one - then putting the while loop in dump_zone instead?

In the conversion to classes, it gets easy to add an iterator using `to_enum()` — one of the motivations to do so.
We can probably extract it to a loop with a yield and a block.
I'll have a play.

mexisme replied:
> I think I'll want to rename this - it sounds like it's just dumping a single period, whereas actually it's dumping everything for the whole zone.

`def dump_periods` ?

jskeet commented on ruby/tzdump.rb:
```diff
@@ -5,35 +7,50 @@
+def puts_crlf(*messages)
+  $stdout.write(messages.join("\r\n"))
+  $stdout.write("\r\n")
+end
```
As noted on previous comments, this doesn't work on Windows - we end up with "\r\r\n" between each line. We definitely need something like the earlier code... it's the only thing I've seen so far that does the right thing on Windows.

jskeet commented on ruby/Gemfile:
```diff
@@ -0,0 +1,3 @@
+source 'https://rubygems.org'
+
+gem 'tzinfo'
```
We need tzinfo-data as well. Presumably bundler will then recreate the gemfile.lock appropriately...

mexisme replied:
Yes.  Just running `bundle` pulls in all the referenced gems.
The caveat is you will probably need to run `bundle exec ./tzdump.rb`.
Hmm, I might add the bundler setup to the file...

mexisme replied:
(Added bundler/setup)

mexisme replied:
Ugh, I misread your comment.  tzinfo-data is not referenced in `tzdump.rb`.
I'll fix that...

mexisme replied:
(Fixed)

Commit:
Refactor into classes and iterators

mexisme commented:
FYI:  I did a further refactor into a more-OO style from the previous more-imperative style.
YMMV, but it "looks" like more-natural Ruby to me.

BTW:  If you're ever likely to want to pass-in more-expressive arguments than just TZ ID's, would probably be good to port to using the Thor gem.

Commit:
Update Gem requirements

Require bundler/setup to stop users having to run `bundle exec $0`
Add tzinfo-data gem to Gemfile

mexisme commented:
No tests, yet   :-/

Commit:
Linter fixes (Rubocop)

Enable sane Rubocop defaults
Fixes to make Rubocop happier

Commit:
Restore copying $stdout FH for console output

jskeet commented on ruby/.rubocop.yml:
```diff
@@ -0,0 +1,16 @@
+# Most of these are disabling existing cops, primarily
```
This all feels like it's getting a bit much for something which is meant to be experimental and short-lived. The metadata living alongside the code is now nearly as big as the code itself!

mexisme replied:
Fair point. I guess I've been enjoying myself too much!

I personally think the code (ignoring all the meta) is pretty solid, though, and I personally feel the structure is very readable and forward-refactorable / portable.

FYI: Rubocop often annoys the XXX out of me...

jskeet commented on ruby/tzdump.rb:
```diff
@@ -1,39 +1,90 @@
-      formatted_start = start.strftime("%Y-%m-%dT%H:%M:%SZ")
-      formatted_offset = format_offset(period.offset.utc_total_offset)
-      console.write "#{formatted_start} #{formatted_offset} #{period.dst? ? "daylight" : "standard"} #{period.offset.abbreviation}\r\n"
+# This is a dumper class for the TZInfo data
```
Then it's not really a zone, is it?

I have to say, these changes look considerably more complicated for no actual benefit. Previously, I understood the code - now I'd be reluctant to change it...

It's definitely interesting from an educational perspective, but I don't think I'll be merging in the PR like this. I may well add the Gemfile and Gemfile.lock and I'll look at taking _some_ of these changes... but I think the overall structure was clearer as it was, given the context of "This is just a small script to dump time zone data, and nothing else."

mexisme replied:
I think it depends on whether you're approaching it as a Rubyist or not.
As a Rubyist, the newer structure is much-more familiar and "conventional" to me.

I totally understand if you're approaching this more from a (e.g.) Perl/Python/Shell _script_ POV;  it looks more-noisy/disconnected than might be your intention.

mexisme replied:
> Then it's not really a zone, is it?

I guessed at a class-name.

jskeet replied:
I'm actually coming at this from a C# and Java perspective, in terms of the OO. It seems odd to build an instance of a class and set properties, just to call a single method - when we can just as easily pass them in as parameters to a static (C#/Java terminology) method. It's not like we _want_ something with state here - we're not going to do anything else afterwards with the combination of the zone and period.

Of course, such a static method could create a stateful object itself if it want to, but I wouldn't have that as the public API - unless I wanted to implement an interface or something similar. (I can imagine creating a "transition formatter" or something similar...)

Just MHO though, and definitely coming from a position of ignorance in Ruby idioms. Maybe creating instances like this is the norm there...

mexisme replied:
I'd say the Ruby idioism is splitting-up large methods into smaller self-explanatory pieces, and since it's super-easy to create classes and attr's: do that instead of passing values around when you don't need to.
I see similarities in Haskell:  avoid "abusing" simple data-types (dicts/hashes) when it's [mostly] painless to create one that encapsulates your intentions and that you can improve iteratively.

I also suspect I've over-engineered a little in a few places (e.g. CrLf class) except I was hoping it would allow more self-description of the intention.

mexisme replied:
If it helps:  some of the conventions and magic take a little getting used-to (e.g. Rails, Capistrano, RSpec) and I personally find languages like Rust more natural and explanatory.

Commit:
Revert to imperative style

Commit:
Linter fixes

Commit:
More Linter fixes

mexisme commented:
OK, I've reverted all the refactor --> classes commit, and rebased the branch to remove some noise.

Commit:
Extract to smaller functions

Commit:
`#id` is deprecated in newer Ruby's, but used to == `#object_id`

mexisme commented:
BTW:  Apologies for all the disjointed replies.  GH wasn't sending them all through to me very consistently, for some reason.

mexisme commented:
Would you prefer I close this PR, @jskeet ?

jskeet commented:
Probably best for the moment, I think. (I'll close now.)

mexisme commented:
Cheers, @jskeet 
Thanks for the time you've spent on it, and I hope it wasn't a waste of yours.
-will

jskeet commented:
Not at all - and I'm sure if this were growing into something bigger, that sort of refactoring would be appropriate. I just find it easier at the moment as it is :)
